server/annotate/cell_type.py

```diff
@@ -0,0 +1,77 @@
+import pickle
+from dataclasses import dataclass
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+# from scvi.models import SCANVI
+from scvi.model import SCANVI
+
+from server.common.utils.data_locator import DataLocator
+
+
+@dataclass
+class CellTypeTissueModel:
+    tissue_type_ontology_term_id: str
+    model_url: str
+    model: SCANVI
+
+# def extract_tissue_types(dataset: AnnData) -> Set[str]:
+#     """
+#     Returns distinct set of tissue type ontology term IDs.
+#     """
+#     pass
+
+
+# @functools.cache
+# def lookup_model(model_repository_base_url: str, tissue_type_ontology_term_id: str) -> CellTypeTissueModel:
+#     model_url = os.path.join(model_repository_base_url, tissue_type_ontology_term_id, ".pkl")
+#     return CellTypeTissueModel(tissue_type_ontology_term_id, model_url=model_url, model=retrieve_model(model_url))
+
+
+def retrieve_model(model_url: str) -> SCANVI:
+    with DataLocator(uri_or_path=model_url).open("rb") as f:
+        model = pickle.load(f)
+        if not isinstance(model, SCANVI):
+            raise "invalid model object at {model_url}: expected type SCANVI, got {model.type}"
+        return model
+
+
+def annotate(query_dataset: AnnData, model: SCANVI, annotation_column_name_prefix: str) -> None:
+    input_dataset = prepare_query_dataset(query_dataset, model)
+
+    # TODO: necessary to partition into AnnData-per-tissue, then invoke model on each partitioned AnnData object,
+    #  then reconstitute into a single AnnData
+    #
+    # TODO: record model metadata (version) in AnnData
+    predictions = model.predict(input_dataset)
+
+    query_dataset.obs[f"{annotation_column_name_prefix}"] = pd.Categorical(predictions)
+    # TODO: How do obtain confidence scores?
```

Comment:
@LuckyMD flagging this for your attention later

Reply:
Once I have some code, I will get to this.

Reply:
Thanks!